### PR TITLE
feat: Add weights parameter to quantile and percentile functions (#32647)

### LINKED ISSUE #32647: Support weighted quantile & percentile
NumPy v2.2.0 added a `weights` parameter to `np.quantile`, `np.percentile`, `np.nanquantile`, and `np.nanpercentile`. We should add support for the `weights` parameter in the associated `jax.numpy` APIs.

### PR DESCRIPTION
**Pull request by C-you-know**
Adds weights parameter to quantile, percentile, nanquantile, and nanpercentile functions for weighted calculations.
Added tests to verify the implementation and ensured existing functionality remains unchanged.
Documentation will be added pending code review.
Happy to make changes and iterate based on feedback!
Fixes #32647

### CODE REVIEW

**Comment by C-you-know** [comment]
Screenshot 2025-12-22 at 4 41 56 PM

Passing all test cases!
Should I add more test cases for edge cases? Let me know if there are specific scenarios you'd like me to cover!

**Comment by C-you-know** [comment]
Gentle ping @jakevdp 🙂

File: jax/_src/numpy/reductions.py [hunk 1]
```diff
@@ -2585,6 +2635,50 @@ def _quantile(a: Array, q: Array, axis: int | tuple[int, ...] | None,
...
+    curr_out_axis = q_ndim
+    for i in range(a.ndim):
+      if i == axis:
+        index.append(idx)
+        if keepdims:
+          curr_out_axis += 1
+      else:
+        index.append(lax.broadcasted_iota(np.int32, out_shape, curr_out_axis))
+        curr_out_axis += 1
+    result = a[tuple(index)]
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: This seems like a very ineficient way to compute the result. I think you could use empty slices in place of the broadcasted iotas to do this more simply.

File: jax/_src/numpy/reductions.py [hunk 2]
```diff
@@ -2585,6 +2635,50 @@ def _quantile(a: Array, q: Array, axis: int | tuple[int, ...] | None,
...
+    cw_p = lax.expand_dims(cum_weights, tuple(range(q_ndim)))
+    is_less = lax.lt(cw_p, target_w_aligned)
+    idx = sum(lax.convert_element_type(is_less, np.int32), axis=axis + q_ndim, keepdims=keepdims)
+    if squash_nans:
+      valid_counts = sum(lax.bitwise_not(nan_data), axis=axis, dtype=q.dtype, keepdims=keepdims)
+    else:
+      valid_counts = lax.full_like(total_weight, a_shape[axis], dtype=q.dtype)
+    limit = lax.sub(valid_counts, lax._const(valid_counts, 1))
+    max_idx = lax.convert_element_type(limit, np.int32)
+    max_idx_p = lax.expand_dims(max_idx, tuple(range(q_ndim)))
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: Style nit: the `_p` suffix in a variable name is typically used for primitives in JAX. It would be clearer to use a different naming convention here. Same with `force_nan_p` below.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks really great on a first read-through, thanks!

A couple small comments below. I'll trigger the tests to see if they reveal any issues with the implementation.

**Comment by C-you-know** [comment]
Hi! I looked into the failing CI cases.

The failures in testWeightedQuantile10 happen during the NumPy reference computation rather than in the JAX implementation itself.

np.nanpercentile emits a RuntimeWarning: All-NaN slice encountered on some randomized inputs; JAX returns the same NaN result but doesn’t emit a warning.

_CheckAgainstNumpy treats the NumPy warning as a test error, so the comparison never reaches the output check.
I believe the implementation is correct, and the issue is how the test handles NumPy warnings. Would you prefer that I update the test to suppress NumPy warnings in the reference path (consistent with other nan* reducer tests), or would you like me to handle this differently?

**Comment by jakevdp** [comment]
JAX tests are deliberately configured to error when there are unexpected warnings. If you want to explicitly ignore a warning within a test, you can use the `jtu.ignore_warning` utility.

File: jax/_src/numpy/reductions.py (hunk 2 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: I have updated the naming in my implementation and replaced the _p suffix with _f to avoid confusion with JAX primitives.

File: jax/_src/numpy/reductions.py (hunk 1 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: I tried using empty slices here but it does not work cleanly because idx already encodes the non-axis dimensions. Combining it with slice(None) causes those dimensions to be included twice under advanced indexing. I used take_along_axis to keep the indexing explicit and avoid that duplication

**Comment by C-you-know** [comment]
Hi @jakevdp, 

I have rebased the PR onto the latest upstream, fixed the failing test cases and made a small naming cleanup in the implementation. All tests pass locally now. 

Thanks for taking another look.

File: jax/_src/numpy/reductions.py [hunk 3]
```diff
@@ -2374,7 +2375,7 @@ def cumulative_prod(
 @export
 @api.jit(static_argnames=('axis', 'overwrite_input', 'keepdims', 'method'))
 def quantile(a: ArrayLike, q: ArrayLike, axis: int | tuple[int, ...] | None = None,
-             out: None = None, overwrite_input: bool = False, method: str = "linear",
+             weights: ArrayLike | None = None, out: None = None, overwrite_input: bool = False, method: str = "linear",
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: `weights` should be a keyword-only parameter, as in `np.quantile`. Same for the other exported functions below.

File: jax/_src/numpy/reductions.py [hunk 4]
```diff
@@ -2462,26 +2478,53 @@ def nanquantile(a: ArrayLike, q: ArrayLike, axis: int | tuple[int, ...] | None =
...
-              method: str, keepdims: bool, squash_nans: bool) -> Array:
-  if method not in ["linear", "lower", "higher", "midpoint", "nearest"]:
-    raise ValueError("method can only be 'linear', 'lower', 'higher', 'midpoint', or 'nearest'")
-  a, = promote_dtypes_inexact(a)
+              method: str, keepdims: bool, squash_nans: bool, weights: ArrayLike | None = None) -> Array:
+  if method not in ["linear", "lower", "higher", "midpoint", "nearest", "inverted_cdf"]:
+    raise ValueError("method can only be 'linear', 'lower', 'higher', 'midpoint', 'nearest' or 'inverted_cdf'")
+  if weights is not None:
+    weights = ensure_arraylike("_quantile", weights)
+    weights = lax.asarray(weights)
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: You can remove this; `ensure_arraylike` already takes care of array conversion. If you move `ensure_arraylike` to the calling function, then you could make the annotation `Array | None` and then this wouldn't be required.

File: jax/_src/numpy/reductions.py [hunk 5]
```diff
@@ -2462,26 +2478,53 @@ def nanquantile(a: ArrayLike, q: ArrayLike, axis: int | tuple[int, ...] | None =
...
 def _quantile(a: Array, q: Array, axis: int | tuple[int, ...] | None,
-              method: str, keepdims: bool, squash_nans: bool) -> Array:
-  if method not in ["linear", "lower", "higher", "midpoint", "nearest"]:
-    raise ValueError("method can only be 'linear', 'lower', 'higher', 'midpoint', or 'nearest'")
-  a, = promote_dtypes_inexact(a)
+              method: str, keepdims: bool, squash_nans: bool, weights: ArrayLike | None = None) -> Array:
+  if method not in ["linear", "lower", "higher", "midpoint", "nearest", "inverted_cdf"]:
+    raise ValueError("method can only be 'linear', 'lower', 'higher', 'midpoint', 'nearest' or 'inverted_cdf'")
+  if weights is not None:
+    weights = ensure_arraylike("_quantile", weights)
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: The name here is used to generate a user-visible message, and `_quantile` is an internal name.

Instead, we should pass weights to `ensure_arraylike` in the calling functions (alternatively we could modify `_quantile` to take the name of the calling function as a parameter, but that seems less clean).

File: jax/_src/numpy/reductions.py [hunk 6]
```diff
@@ -2462,26 +2478,53 @@ def nanquantile(a: ArrayLike, q: ArrayLike, axis: int | tuple[int, ...] | None =
...
+    weights = lax.asarray(weights)
+    if dtypes.issubdtype(weights.dtype, np.complexfloating):
+      raise ValueError("Weights cannot be complex types.")
+    if method != "inverted_cdf":
+      raise NotImplementedError(f"{method} doesn't support weights. Only method 'inverted_cdf' supports weights.")
+    a, weights = promote_dtypes_inexact(a, weights)
+    if weights.shape != a.shape:
+      if axis is None:
+        raise ValueError("Weights shape must match 'a' shape when axis is None.")
+      ax_tuple = (axis,) if isinstance(axis, int) else tuple(axis)
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: You can use `util.canonicalize_axis_tuple` for this.

File: jax/_src/numpy/reductions.py [hunk 7]
```diff
@@ -2462,26 +2478,53 @@ def nanquantile(a: ArrayLike, q: ArrayLike, axis: int | tuple[int, ...] | None =
...
+    a, = promote_dtypes_inexact(a)
   keepdim = []
   if dtypes.issubdtype(a.dtype, np.complexfloating):
     raise ValueError("quantile does not support complex input, as the operation is poorly defined.")
   if axis is None:
     if keepdims:
       keepdim = [1] * a.ndim
     a = a.ravel()
+    if weights is not None:
+        weights = weights.ravel()
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: two-space indentation please. Also in a few places below.

File: jax/_src/numpy/reductions.py [hunk 8]
```diff
@@ -2585,6 +2636,57 @@ def _quantile(a: Array, q: Array, axis: int | tuple[int, ...] | None,
     result = lax.select(pred, low_value, high_value)
   elif method == "midpoint":
     result = lax.mul(lax.add(low_value, high_value), lax._const(low_value, 0.5))
+  elif method == "inverted_cdf":
+    if weights is None:
+      weights = lax.full(a.shape, 1.0, dtype=a.dtype)
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: `lax.full_like(a, 1.0)` is a bit more concise, and will propagate sharding information when relevant.

File: jax/_src/numpy/reductions.py [hunk 9]
```diff
@@ -2585,6 +2636,57 @@ def _quantile(a: Array, q: Array, axis: int | tuple[int, ...] | None,
...
+    target_w_aligned = target_w if keepdims else lax.expand_dims(target_w, (axis + q_ndim,))
+    cw_f = lax.expand_dims(cum_weights, tuple(range(q_ndim)))
+    is_less = lax.lt(cw_f, target_w_aligned)
+    idx = sum(lax.convert_element_type(is_less, np.int32), axis=axis + q_ndim, keepdims=keepdims)
+    if squash_nans:
+      valid_counts = sum(lax.bitwise_not(nan_data), axis=axis, dtype=q.dtype, keepdims=keepdims)
+    else:
+      valid_counts = lax.full_like(total_weight, a_shape[axis], dtype=q.dtype)
+    limit = lax.sub(valid_counts, lax._const(valid_counts, 1))
+    max_idx = lax.convert_element_type(limit, np.int32)
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: Let's use `dtypes.default_int_dtype()` here.

File: jax/_src/numpy/reductions.py [hunk 10]
```diff
@@ -2585,6 +2636,57 @@ def _quantile(a: Array, q: Array, axis: int | tuple[int, ...] | None,
...
+    max_idx_f = lax.expand_dims(max_idx, tuple(range(q_ndim)))
+    max_idx_f = lax.convert_element_type(max_idx_f, idx.dtype)
+    idx = lax.max(lax._const(idx, 0), lax.min(idx, max_idx_f))
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py:2660]: You could use `lax.clamp` here.

File: jax/_src/numpy/reductions.py [hunk 11]
```diff
@@ -2585,6 +2636,57 @@ def _quantile(a: Array, q: Array, axis: int | tuple[int, ...] | None,
...
+    if keepdims:
+      idx_take = lax.squeeze(idx, (q_ndim + axis,))
+    else:
+      idx_take = idx
+    if q_ndim == 0:
+      idx_transposed = lax.expand_dims(idx_take, (axis,))
+      result = indexing.take_along_axis(a, idx_transposed, axis=axis)
+      result = lax.squeeze(result, (axis,))
+    else:
+      perm = (list(range(q_ndim, q_ndim + axis)) +
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: Generalized unpacking would be a bit clearer here I think:
```
perm = [*range(q_ndim, q_ndim + axis),
        *range(q_ndim),
        *range(q_ndim + axis, idx_take.ndim)]
```

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looking good! A few more changes below.

File: jax/_src/numpy/reductions.py (hunk 3 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: Ok. Updated.

File: jax/_src/numpy/reductions.py (hunk 5 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: Updated accordingly

File: jax/_src/numpy/reductions.py (hunk 6 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: Updated accordingly

File: jax/_src/numpy/reductions.py (hunk 7 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: Sorry. Updated the code to follow two space indentation.

File: jax/_src/numpy/reductions.py (hunk 8 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: Updated

File: jax/_src/numpy/reductions.py (hunk 9 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: OK. Updated.

File: jax/_src/numpy/reductions.py (hunk 11 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: OK. Updated.

File: jax/_src/numpy/reductions.py (hunk 10 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py:2660]: I considered using lax.clamp here but it would require dynamically broadcasting max_idx_f to match the shape of idx. I kept the current approach to avoid introducing that additional broadcasting logic. I am happy to switch to lax.clamp if you would prefer that for consistency

File: jax/_src/numpy/reductions.py (hunk 4 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: I wanted to clarify this point based on the current structure of quantile and the related nanquantile, percentile and nanpercentile functions. The calling functions already apply ensure_arraylike to an and q but still pass both through lax.asarray before calling _quantile. since, ensure_arraylike also performs array conversion I was not sure whether this duplication is intentional for consistency.

**Comment by C-you-know** [comment]
I have addressed the review comments and left inline replies where relevant. Please let me know if I missed anything

**Comment by C-you-know** [comment]
Hi @jakevdp,
Following up on this PR. I addressed the review comments and pushed updates last week. Please let me know if there is anything else you would like me to revise.

Thanks again for your time reviewing this.

File: jax/_src/numpy/reductions.py (hunk 4 above)
Comment (jakevdp) [reply jax/_src/numpy/reductions.py]: `lax.asarray` is not necessary for any value that's already passed through `ensure_arraylike`.

**Comment by C-you-know** [comment]
Hi @jakevdp, thanks for the feedback! I have removed the redundant lax.asarray call as suggested - ensure_arraylike already handles the conversion so it was not needed. I have also updated the code to use lax.convert_element_type where dtype casting is required, following the latest internal API changes. All 680 tests are passing. Let me know if there is anything else to address!

File: jax/_src/numpy/reductions.py [hunk 12]
```diff
@@ -2407,19 +2412,28 @@ def quantile(a: ArrayLike, q: ArrayLike, axis: int | tuple[int, ...] | None = No
...
+    Array(3., dtype=float32)
   """
   a, q = ensure_arraylike("quantile", a, q)
+  if weights is not None:
+    weights = ensure_arraylike("quantile", weights) 
   if overwrite_input or out is not None:
     raise ValueError("jax.numpy.quantile does not support overwrite_input=True "
                      "or out != None")
-  return _quantile(a, q, axis, method, keepdims, False)
+  return _quantile(lax.asarray(a), lax.asarray(q), axis, method, keepdims, False, weights)
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: `lax.asarray` should be removed in both places here: it's redundant because of the `ensure_arraylike` call above.

File: jax/_src/numpy/reductions.py [hunk 13]
```diff
@@ -2457,26 +2476,52 @@ def nanquantile(a: ArrayLike, q: ArrayLike, axis: int | tuple[int, ...] | None =
...
   """
   a, q = ensure_arraylike("nanquantile", a, q)
+  if weights is not None:
+    weights = ensure_arraylike("nanquantile", weights)
   if overwrite_input or out is not None:
     msg = ("jax.numpy.nanquantile does not support overwrite_input=True or "
            "out != None")
     raise ValueError(msg)
-  return _quantile(a, q, axis, method, keepdims, True)
+  return _quantile(lax.asarray(a), lax.asarray(q), axis, method, keepdims, True, weights)
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: Remove `lax.asarray` here as well.

**Review by jakevdp — COMMENTED** [review]
Looks good, just a couple more unnecessary `lax.asarray` calls to remove.

File: jax/_src/numpy/reductions.py (hunk 12 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: updated.

File: jax/_src/numpy/reductions.py (hunk 13 above)
Comment (C-you-know) [reply jax/_src/numpy/reductions.py]: updated.

**Comment by C-you-know** [comment]
Hi @jakevdp, I have removed all the unnecessary lax.asarray calls. Please take a look!

**Comment by jakevdp** [comment]
Looks good, just a couple things:

- there's some trailing whitespace lint
- changes need to be squashed into a single commit

**Comment by C-you-know** [comment]
Hi @jakevdp, I have fixed the trailing whitespace and squashed all commits into a single one. Please take a look!

**Review by jakevdp — APPROVED** [review]
